### ISSUE TITLE
Find different ways of representing learning

### ISSUE THREAD

**Issue by jsueling** [rl-squad/planning#11]
# Saliency maps/Feature importance

The gradients of the models output wrt. its input features

### The rough idea (needs refining/further thought)

The saliency/gradient represents how small changes in the state input affect the Q value prediction. Low abs grads means our QNet/critic does not care about those state dims. Could fix a state and action chosen by the policy, replay the saliency test with same (s, a) at different points of training. Graphically represent wrt the 17 state dims and interpret results. Which (s, a) to choose?

### Example code
```python
import torch
# Assume a trained policy network `model` and a state `x` (tensor)
x = torch.tensor(walker2d_state, requires_grad=True)
# action is deterministic in DDPG/TD3
action = actor_network(x)  # [a1, a2, ...]
q_value = critic(x, action)
# Saliency for a1 (how inputs affect hip torque decision)
q_value[0].backward(retain_graph=True)
# 17 grads that show how much does perturbing each state feature affect the Q-value for the hip_torque action
saliency_hip = x.grad.abs()
x.grad.zero_()
# Saliency for a2 (knee torque)
q_value[1].backward()
saliency_knee = x.grad.abs(retain_graph=True)

# Or sample actions, perhaps at the start of the episode and take the mean 
q_values = [critic(x, actor(x)) for _ in range(10)]  # Sample 10 actions
q_avg = torch.stack(q_values).mean()  # E[Q(s, a)]
q_avg.backward()
...
```
